mmdet/models/layers/transformer/dab_detr_transformer.py

```diff
@@ -0,0 +1,300 @@
+        """
+        output = query
+        unsigmoid_references = query_pos
+
+        reference_points = unsigmoid_references.sigmoid()
+        intermediate_reference_points = [reference_points]
+
+        intermediate = []
+        for layer_id, layer in enumerate(self.layers):
+            obj_center = reference_points[..., :self.query_dim]
+            ref_sine_embed = convert_coordinate_to_encoding(
+                coord_tensor=obj_center, num_feats=self.embed_dims // 2)
+            query_pos = self.ref_point_head(
+                ref_sine_embed)  # [bs, nq, 2c] -> [bs, nq, c]
+            # For the first decoder layer, do not apply transformation
+            if self.query_scale_type != 'fix_elewise':
+                if layer_id == 0:
+                    pos_transformation = 1
+                else:
+                    pos_transformation = self.query_scale(output)
+            else:
+                pos_transformation = self.query_scale.weight[layer_id]
+            # apply transformation
+            ref_sine_embed = ref_sine_embed[
+                ..., :self.embed_dims] * pos_transformation
+            # modulated height and weight attention
+            if self.with_modulated_hw_attn:
+                assert obj_center.size(-1) == 4
+                ref_hw = self.ref_anchor_head(output).sigmoid()
+                ref_sine_embed[..., self.embed_dims // 2:] *= \
+                    (ref_hw[..., 0] / obj_center[..., 2]).unsqueeze(-1)
+                ref_sine_embed[..., : self.embed_dims // 2] *= \
+                    (ref_hw[..., 1] / obj_center[..., 3]).unsqueeze(-1)
+
+            output = layer(
+                output,
+                key,
+                query_pos=query_pos,
+                ref_sine_embed=ref_sine_embed,
+                key_pos=key_pos,
+                key_padding_mask=key_padding_mask,
+                is_first=(layer_id == 0),
+                **kwargs)
+            # iter update
+            tmp_reg_preds = reg_branches(output)
+            tmp_reg_preds[..., :self.query_dim] += inverse_sigmoid(
+                reference_points)
+            new_reference_points = tmp_reg_preds[
+                ..., :self.query_dim].sigmoid()
+            if layer_id != self.num_layers - 1:
+                intermediate_reference_points.append(new_reference_points)
+            reference_points = new_reference_points.detach()
+
+            if self.return_intermediate:
+                intermediate.append(self.post_norm(output))
+
+        output = self.post_norm(output)
+
+        if self.return_intermediate:
+            return [
+                torch.stack(intermediate),
+                torch.stack(intermediate_reference_points),
+            ]
+        else:
+            return [
+                output.unsqueeze(0),
+                torch.stack(intermediate_reference_points)
+            ]
+
+
+class DabDetrTransformerEncoder(DetrTransformerEncoder):
```

Comment:
DAB may be better because it represents the `Dynamic Anchor Boxes `

Reply:
ok 👍 :)